src/main/java/io/cdap/plugin/MongoDBConfig.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin;
+
+import com.google.common.base.Strings;
+import io.cdap.cdap.api.annotation.Description;
+import io.cdap.cdap.api.annotation.Macro;
+import io.cdap.cdap.api.annotation.Name;
+import io.cdap.cdap.api.plugin.PluginConfig;
+import io.cdap.cdap.etl.api.validation.InvalidConfigPropertyException;
+import io.cdap.plugin.common.Constants;
+import io.cdap.plugin.common.IdUtils;
+
+import javax.annotation.Nullable;
+
+/**
+ * Defines a base {@link PluginConfig} that MongoDB Source and Sink can re-use.
+ */
+public class MongoDBConfig extends PluginConfig {
+
+  @Name(Constants.Reference.REFERENCE_NAME)
+  @Description(Constants.Reference.REFERENCE_NAME_DESCRIPTION)
+  @Macro
+  public String referenceName;
+
+  @Name(MongoDBConstants.HOST)
+  @Description("Host that MongoDB is running on.")
+  @Macro
+  public String host;
+
+  @Name(MongoDBConstants.PORT)
+  @Description("Port that MongoDB is listening to.")
+  @Macro
+  public Integer port;
+
+  @Name(MongoDBConstants.DATABASE)
+  @Description("MongoDB database name.")
+  @Macro
+  public String database;
+
+  @Name(MongoDBConstants.COLLECTION)
+  @Description("Name of the database collection.")
+  @Macro
+  public String collection;
+
+  @Name(MongoDBConstants.USER)
+  @Description("User to use to connect to the specified database. Required for databases that " +
+    "need authentication. Optional for databases that do not require authentication.")
+  @Macro
+  @Nullable
+  public String user;
+
+  @Name(MongoDBConstants.PASSWORD)
+  @Description("Password to use to connect to the specified database. Required for databases that " +
+    "need authentication. Optional for databases that do not require authentication.")
+  @Macro
+  @Nullable
+  public String password;
+
+  @Name(MongoDBConstants.CONNECTION_ARGUMENTS)
+  @Description("A list of arbitrary string key/value pairs as connection arguments.")
+  @Macro
+  @Nullable
+  public String connectionArguments;
+
+  /**
+   * Validates {@link MongoDBConfig} instance.
+   */
+  public void validate() {
+    if (!containsMacro(Constants.Reference.REFERENCE_NAME) && Strings.isNullOrEmpty(referenceName)) {
+      throw new InvalidConfigPropertyException("Reference name must be specified", Constants.Reference.REFERENCE_NAME);
+    } else {
```

Comment:
[Changed to](https://github.com/vladhlinskiy/mongodb-plugins/pull/1/files#diff-2e06cbe41ef929c2505616b958113f83R84):
```
    if (!containsMacro(Constants.Reference.REFERENCE_NAME) && Strings.isNullOrEmpty(referenceName)) {
      throw new InvalidConfigPropertyException("Reference name must be specified", Constants.Reference.REFERENCE_NAME);
    } else if (!containsMacro(Constants.Reference.REFERENCE_NAME)) {
      try {
        IdUtils.validateId(referenceName);
      } catch (IllegalArgumentException e) {
        // InvalidConfigPropertyException should be thrown instead of IllegalArgumentException
        throw new InvalidConfigPropertyException("Invalid reference name", e, Constants.Reference.REFERENCE_NAME);
      }
    }
```